File: python/src/isValid.py
# https://leetcode.com/problems/valid-parentheses/


from collections import deque


class Solution:
    def isValid(self, s: str) -> bool:
        q = deque()
        pairs = {
            "(": ")",
            "[": "]",
            "{": "}",
        }

        for c in s:
            if c in pairs.keys():
                q.append(pairs[c])
                continue

            if len(q) == 0:
                return False

            if q.pop() != c:
                return False

        if len(q) != 0:
            return False

        return True

File: go/isValid.go
// https://leetcode.com/problems/valid-parentheses

package leetcode

func isValid(s string) bool {
	if len(s)%2 == 1 {
		return false
	}

	var stack []rune
	closingParentheses := map[rune]rune{
		'(': ')',
		'{': '}',
		'[': ']',
	}

	for _, val := range s {
		if closing, ok := closingParentheses[val]; ok {
			stack = append(stack, closing)
			continue
		}

		if len(stack) == 0 {
			return false
		}

		if val != stack[len(stack)-1] {
			return false
		}

		stack = stack[:len(stack)-1]
	}

	return len(stack) == 0
}
